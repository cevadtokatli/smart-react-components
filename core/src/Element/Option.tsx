import styled from 'styled-components'
import applyStyledProps, { StyledProps, shouldForwardProp } from '../styled-props'

export default styled.option.withConfig({ shouldForwardProp })<StyledProps>(applyStyledProps)
