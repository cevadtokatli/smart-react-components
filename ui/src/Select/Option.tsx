import { ContentElement, PaletteProp, SetState } from '@smart-react-components/core/types'
import React from 'react'
import DropdownItemElement from '../components/Dropdown/DropdownItemElement'
import { FormValue } from '../types'
import WaveEffect from '../WaveEffect'

export interface Props {
  children: ContentElement
  isDisabled?: boolean
  value: FormValue
}

export interface PrivateProps {
  active: FormValue | FormValue[]
  hasHover: boolean
  hasWaveEffect: boolean
  isDisabled: boolean
  isEmbedded: boolean
  isOutline: boolean
  isSoft: boolean
  palette: PaletteProp
  setActive: SetState<FormValue | FormValue[]>
  waveEffectPalette: PaletteProp
}

const Option: React.FC<Props> = ({ active, children, hasHover, hasWaveEffect, isDisabled, isOutline, isSoft, palette, setActive, value, waveEffectPalette }: (Props & PrivateProps)) => {
  const handleOnClick = () => {
    if (isDisabled) {
      return
    }

    if (!Array.isArray(active)) {
      setActive(value)
    } else {
      if (!active.includes(value)) {
        setActive([...active, value])
      } else {
        setActive(active.filter(i => i !== value))
      }
    }
  }

  const isActive = () => !Array.isArray(active) ? value === active : active.includes(value)

  const content = (
    <DropdownItemElement
      hasHover={hasHover}
      isActive={isActive()}
      isDisabled={isDisabled}
      isOutline={isOutline}
      isSoft={isSoft}
      palette={palette}
      onClick={handleOnClick}
    >
      {children}
    </DropdownItemElement>
  )

  return hasWaveEffect ? <WaveEffect palette={waveEffectPalette}>{content}</WaveEffect> : content
}

export default Option
