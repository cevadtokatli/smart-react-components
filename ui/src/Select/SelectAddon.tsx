import InputAddon from '../Input/InputAddon'

export default InputAddon
