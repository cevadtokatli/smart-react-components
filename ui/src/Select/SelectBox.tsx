import Div from '@smart-react-components/core/Element/Div'
import { GenericProps } from '../types/form'
import extractElementProps from '@smart-react-components/core/element-props'
import intrinsicStyledCoreProps from '@smart-react-components/core/element-props/intrinsic-styled-core-props'
import useChangeEffect from '@smart-react-components/core/hooks/useChangeEffect'
import { Theme } from '@smart-react-components/core/theme'
import { ContentElement, JSXElementProps } from '@smart-react-components/core/types'
import React from 'react'
import { useTheme } from 'styled-components'
import Badge from '../Badge'
import BadgeIcon from '../Badge/BadgeIcon'
import BadgeList from '../Badge/BadgeList'
import FormBlockLabel from '../components/Form/FormBlockLabel'
import HiddenInput from '../components/Form/HiddenInput'
import InputWrapper from '../components/Input/InputWrapper'
import Dropdown from '../Dropdown'
import DropdownArrowIcon from '../Dropdown/DropdownArrowIcon'
import DropdownListElement from '../components/Dropdown/DropdownListElement'
import useAddons from '../hooks/useAddons'
import useSelectBoxItemList from '../hooks/useSelectBoxItemList'
import CloseIcon from '../icons/Close'
import InputElement from '../components/Input/InputElement'
import InputPlaceholder from '../components/Input/InputPlaceholder'
import useSelectBoxHover from '../hooks/useSelectBoxHover'
import { FormValue } from '../types'
import { getInputValue } from '../util/form'
import { applyResponsiveStyledProp } from '../util/props'
import { getWaveEffectPalette } from '../util/wave-effect'
import InputAddon from './SelectAddon'

export interface Props extends GenericProps {
  dropdownArrowIconElementProps?: JSXElementProps
  placeholder?: string
}

const SelectBox = React.forwardRef<HTMLInputElement, Props>((props, forwardRef) => {
  const theme = useTheme() as Theme

  const formEl = React.useRef<HTMLDivElement>(null)

  const { itemList } = useSelectBoxItemList({ children: props.children })

  const handleBadgeClick = value => props.setActive((props.active as FormValue[]).filter(i => i !== value))

  const content = React.useMemo<ContentElement>(() => {
    if (
      typeof props.active === 'undefined'
      || props.active === null
      || (Array.isArray(props.active) && props.active.length === 0)
      || (props.active === '' && !itemList[''])
    ) {
      return <InputPlaceholder>{props.placeholder}</InputPlaceholder>
    }

    if (!Array.isArray(props.active)) {
      return <div>{itemList[String(props.active)]?.children}</div>
    }

    return (
      <BadgeList
        palette={props.palette}
        isSoft={props.isSoft}
        shape={props.shape}
        size={props.size}
        sizeSm={props.sizeSm}
        sizeMd={props.sizeMd}
        sizeLg={props.sizeLg}
        sizeXl={props.sizeXl}
      >
        {
          props.active
            .map(item => item)
            .sort((a, b) => itemList[String(a)].idx < itemList[String(b)].idx ? -1 : 1)
            .map(item => (
              <Badge key={String(item)} elementProps={{ 'data-src-not-clickable': true }}>
                <BadgeIcon onClick={() => handleBadgeClick(item)}>
                  <CloseIcon />
                </BadgeIcon>
                {itemList[String(item)].children}
              </Badge>
            ))
        }
      </BadgeList>
    )
  }, [props.active, props.placeholder, itemList])

  const [dropdownStatus, setDropdownStatus] = React.useState(false)

  const waveEffectPalette = React.useMemo(() => getWaveEffectPalette(props, theme.$.vars.isDarkMode), [props.waveEffectPalette, props.palette, props.isOutline, props.isSoft, theme.$.vars.isDarkMode])

  const { hovered, setHovered } = useSelectBoxHover({
    active: props.active,
    children: Array.isArray(props.children) ? props.children : [props.children],
    dropdownStatus,
    hasHover: props.hasHover,
    isDisabled: props.isDisabled,
    setActive: props.setActive,
    setDropdownStatus,
  })

  useChangeEffect(() => {
    const event = new Event('src.fixedBox.setPosition')
    formEl.current.dispatchEvent(event)
  }, [content])

  const { leftAddon, rightAddon } = useAddons({
    Component: InputAddon,
    leftAddon: props.leftAddon,
    rightAddon: props.rightAddon,
    props: {
      cursorKey: 'selectBox',
      hasBorder: props.hasBorder,
      isDisabled: props.isDisabled,
      isFocused: false,
      isOutline: props.isOutline,
      isSoft: props.isSoft,
      palette: props.palette,
      shape: props.shape,
      size: props.size,
      sizeSm: props.sizeSm,
      sizeMd: props.sizeMd,
      sizeLg: props.sizeLg,
      sizeXl: props.sizeXl,
    },
  })

  const handleOptionClick = active => {
    props.setActive(active)

    if (!Array.isArray(active)) {
      setDropdownStatus(false)
    }
  }

  return (
    <Dropdown
      elementProps={props.elementProps}
      status={dropdownStatus}
      setStatus={setDropdownStatus}
    >
      <FormBlockLabel
        {...extractElementProps(props, [intrinsicStyledCoreProps])}
        as="div"
        cursorKey="selectBox"
        formSize={props.size}
        formSizeSm={props.sizeSm}
        formSizeMd={props.sizeMd}
        formSizeLg={props.sizeLg}
        formSizeXl={props.sizeXl}
        isBlock={props.isBlock}
        isDisabled={props.isDisabled}
        ref={formEl}
      >
        { props.label && <span>{props.label}</span> }
        <Div
          display="flex"
          flex="1 1 auto"
        >
          { leftAddon?.props?.isSeparated && leftAddon }
          <InputWrapper
            cursor="$cursor.selectBox"
            hasBorder={props.hasBorder}
            hasSeparatedLeftAddon={leftAddon?.props?.isSeparated}
            hasSeparatedRightAddon={rightAddon?.props?.isSeparated}
            isDisabled={props.isDisabled}
            isFocused={false}
            isOutline={props.isOutline}
            isSoft={props.isSoft}
            palette={props.palette}
            shape={props.shape}
          >
            { (leftAddon && !leftAddon?.props?.isSeparated) && leftAddon }
            <InputElement
              hasLeftAddon={!!leftAddon}
              hasRightAddon={!!rightAddon}
              inputSize={props.size}
              inputSizeSm={props.sizeSm}
              inputSizeMd={props.sizeMd}
              inputSizeLg={props.sizeLg}
              inputSizeXl={props.sizeXl}
              isDisabled={props.isDisabled}
              isFocused={false}
              isInput={false}
              isOutline={props.isOutline}
              isSoft={props.isSoft}
              palette={props.palette}
            >
              <HiddenInput
                onChange={() => {}}
                ref={forwardRef}
                value={getInputValue(props.active)}
                {...(props.isDisabled && { disabled: true })}
                {...(props.isRequired && { required: true })}
              />
              { content }
              <DropdownArrowIcon
                fill="currentcolor"
                {...applyResponsiveStyledProp(props, 'size', 'iconSize', v => `$size.icon.${v}`)}
                {...props.dropdownArrowIconElementProps}
              />
            </InputElement>
            { (rightAddon && !rightAddon?.props?.isSeparated) && rightAddon }
          </InputWrapper>
          { rightAddon?.props?.isSeparated && rightAddon }
        </Div>
      </FormBlockLabel>
      <DropdownListElement
        dropdownListSize={props.size}
        dropdownListSizeSm={props.sizeSm}
        dropdownListSizeMd={props.sizeMd}
        dropdownListSizeLg={props.sizeLg}
        dropdownListSizeXl={props.sizeXl}
        isOutline={props.isOutline}
        isSoft={props.isSoft}
        palette={props.palette}
      >
        { (Array.isArray(props.children) ? props.children : [props.children]).map((item, idx) => item && React.cloneElement(item, {
          key: item.key ?? idx,
          active: props.active,
          cursorKey: 'selectBox',
          hasHover: false,
          hasWaveEffect: props.hasWaveEffect,
          hovered,
          isEmbedded: false,
          isOutline: props.isOutline,
          isSoft: props.isSoft,
          palette: props.palette,
          setActive: handleOptionClick,
          setHovered: props.hasHover ? setHovered : null,
          waveEffectPalette,
          ...(props.isDisabled && { isDisabled: true }),
        })) }
      </DropdownListElement>
    </Dropdown>
  )
})

SelectBox.defaultProps = {
  dropdownArrowIconElementProps: {},
  elementProps: {},
  hasBorder: true,
  hasHover: true,
  hasWaveEffect: true,
  isBlock: true,
  isOutline: true,
  palette: 'primary',
  shape: 'rectangle',
  size: 'medium',
}

export default SelectBox
