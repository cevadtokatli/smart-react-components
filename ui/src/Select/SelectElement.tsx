import styled from 'styled-components'
import InputElement, { Props } from '../Input/InputElement'
import { toCSSValue } from '../util/css'

export default styled(InputElement).attrs<Props>({
  as: 'div',
  getInputSize: (v, t) => `
    font-size: ${t.$.size.input[v].fontSize};

    > div {
      margin: ${toCSSValue(t.$.size.input[v].padding.y)(v => v / 2)} ${toCSSValue(t.$.size.input[v].padding.x)(v => v / 2)};
      padding: ${toCSSValue(t.$.size.input[v].padding.y)(v => v / 2)} ${toCSSValue(t.$.size.input[v].padding.x)(v => v / 2)};
    }
  `,
  isInput: false,
})<Props>``
