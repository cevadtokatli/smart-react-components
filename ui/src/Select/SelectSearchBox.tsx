import extractElementProps from '@smart-react-components/core/element-props'
import changeEvents, { ChangeEvents } from '@smart-react-components/core/element-props/change-events'
import focusEvents, { FocusEvents } from '@smart-react-components/core/element-props/focus-events'
import { IntrinsicStyledCoreProps } from '@smart-react-components/core/element-props/intrinsic-styled-core-props'
import intrinsicStyledFlexProps from '@smart-react-components/core/element-props/intrinsic-styled-flex-props'
import intrinsicStyledMarginProps from '@smart-react-components/core/element-props/intrinsic-styled-margin-props'
import intrinsicStyledSizeProps from '@smart-react-components/core/element-props/intrinsic-styled-size-props'
import keyboardEvents, { KeyboardEvents } from '@smart-react-components/core/element-props/keyboard-events'
import useChangeEffect from '@smart-react-components/core/hooks/useChangeEffect'
import { Theme } from '@smart-react-components/core/theme'
import { ContentElement, JSXChildren, JSXElementProps, PaletteProp, ResponsiveProp, SetState, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import { useTheme } from 'styled-components'
import DropdownListElement from '../components/Dropdown/DropdownListElement'
import useSelectBoxHover from '../hooks/useSelectBoxHover'
import { getWaveEffectPalette } from '../util/wave-effect'
import Dropdown from '../Dropdown'
import Input from '../Input'

export interface Props extends
  Partial<ResponsiveProp<'size', SizeProp>>,
  IntrinsicStyledCoreProps,
  ChangeEvents,
  FocusEvents,
  KeyboardEvents {
  children: JSXChildren
  containerProps?: JSXElementProps
  dropdownProps?: JSXElementProps
  hasBorder?: boolean
  hasHover?: boolean
  hasWaveEffect?: boolean
  hasSpellCheck?: boolean
  inputProps?: JSXElementProps
  isBlock?: boolean
  isDisabled?: boolean
  isOutline?: boolean
  isReadOnly?: boolean
  isRequired?: boolean
  isSoft?: boolean
  label?: ContentElement
  leftAddon?: ContentElement
  palette?: PaletteProp
  placeholder?: string
  rightAddon?: ContentElement
  setValue: SetState<string>
  shape?: ShapeProp
  template?: JSX.Element
  value: string
  waveEffectPalette?: PaletteProp
}

const SelectSearchBox = React.forwardRef<HTMLInputElement, Props>((props, forwardRef) => {
  const theme = useTheme() as Theme

  const [dropdownStatus, setDropdownStatus] = React.useState(false)

  const lastOptionList = React.useRef<JSX.Element[]>([])

  const applySearch = React.useCallback((children: JSX.Element[], list: JSX.Element[], value: string) => {
    children.forEach(item => {
      switch (item.type.displayName) {
        case 'Option':
          if (!value || item.props.value.trim().toLowerCase().includes(value)) {
            list.push(item)
          }
          break
        case 'OptionGroup':
          const group = React.cloneElement(item, { children: [] })
          if (item.props.children) {
            applySearch(Array.isArray(item.props.children) ? item.props.children : [item.props.children], group.props.children, value)
          }
          if (group.props.children.length > 0) {
            list.push(group)
          }
          break
      }
    })
  }, [])

  const optionList = React.useMemo<JSX.Element[]>(() => {
    /**
     * Returns the same list when the dropdown gets closed since it is not needed to update during the closing animation.
     */
    if (!dropdownStatus && lastOptionList.current.length > 0) {
      return lastOptionList.current
    }

    const children = []
    const value = props.value.trim().toLowerCase()
    applySearch(Array.isArray(props.children) ? props.children : [props.children], children, value)

    if (dropdownStatus) {
      lastOptionList.current = children
    }

    return children
  }, [props.children, props.value, dropdownStatus])

  const { hovered, setHovered } = useSelectBoxHover({
    active: props.value,
    children: optionList,
    dropdownStatus,
    hasHover: props.hasHover,
    setActive: props.setValue,
    setDropdownStatus,
  })

  const waveEffectPalette = React.useMemo(() => getWaveEffectPalette(props, theme.$.vars.isDarkMode), [props.waveEffectPalette, props.palette, props.isOutline, props.isSoft, theme.$.vars.isDarkMode])

  useChangeEffect(() => {
    setHovered(undefined)
  }, [optionList])

  const handleOptionClick = value => {
    setDropdownStatus(false)
    props.setValue(value)
  }

  return (
    <Dropdown
      elementProps={props.dropdownProps}
      status={dropdownStatus}
      setStatus={setDropdownStatus}
    >
      <Input
        {...extractElementProps(props, [changeEvents, focusEvents, intrinsicStyledFlexProps, intrinsicStyledMarginProps, intrinsicStyledSizeProps, keyboardEvents])}
        containerProps={props.containerProps}
        hasBorder={props.hasBorder}
        hasSpellCheck={props.hasSpellCheck}
        inputProps={props.inputProps}
        isBlock={props.isBlock}
        isDisabled={props.isDisabled}
        isOutline={props.isOutline}
        isReadOnly={props.isReadOnly}
        isRequired={props.isRequired}
        isSoft={props.isSoft}
        label={props.label}
        leftAddon={props.leftAddon}
        palette={props.palette}
        placeholder={props.placeholder}
        ref={forwardRef}
        rightAddon={props.rightAddon}
        setValue={props.setValue}
        shape={props.shape}
        size={props.size}
        sizeSm={props.sizeSm}
        sizeMd={props.sizeMd}
        sizeLg={props.sizeLg}
        sizeXl={props.sizeXl}
        template={props.template}
        value={props.value}
      />
      <DropdownListElement
        dropdownListSize={props.size}
        dropdownListSizeSm={props.sizeSm}
        dropdownListSizeMd={props.sizeMd}
        dropdownListSizeLg={props.sizeLg}
        dropdownListSizeXl={props.sizeXl}
        isOutline={props.isOutline}
        isSoft={props.isSoft}
        palette={props.palette}
      >
        { optionList.map((item, idx) => item && React.cloneElement(item, {
          key: item.key ?? idx,
          active: props.value,
          cursorKey: 'selectBox',
          hasHover: false,
          hasWaveEffect: props.hasWaveEffect,
          hovered,
          isEmbedded: false,
          isOutline: props.isOutline,
          isSoft: props.isSoft,
          palette: props.palette,
          setActive: handleOptionClick,
          setHovered: props.hasHover ? setHovered : null,
          waveEffectPalette,
          ...(props.isDisabled && { isDisabled: true }),
        })) }
      </DropdownListElement>
    </Dropdown>
  )
})

SelectSearchBox.defaultProps = {
  containerProps: {},
  dropdownProps: {},
  hasHover: true,
  hasWaveEffect: true,
  isOutline: true,
  palette: 'primary',
  size: 'medium',
}

export default SelectSearchBox
