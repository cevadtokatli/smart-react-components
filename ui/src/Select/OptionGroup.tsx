import Div from '@smart-react-components/core/Element/Div'
import { ContentElement } from '@smart-react-components/core/types'
import React from 'react'
import { PrivateProps } from './Option'

export interface Props {
  children: JSX.Element | JSX.Element[]
  label: ContentElement
}

const OptionGroup: React.FC<Props> = ({ children, active, hasHover, hasWaveEffect, isDisabled, isEmbedded, isOutline, isSoft, label, palette, setActive, waveEffectPalette }: Props & PrivateProps) => (
  <>
    <Div fontWeight="$fontWeight.semibold">{label}</Div>
    { (Array.isArray(children) ? children : [children]).map((item, idx) => React.cloneElement(item, {
      key: item.key ?? idx,
      active,
      hasHover,
      hasWaveEffect,
      isEmbedded,
      isOutline,
      isSoft,
      palette,
      setActive,
      waveEffectPalette,
      ...(isDisabled && { isDisabled }),
    })) }
  </>
)

OptionGroup.displayName = 'OptionGroup'

export default OptionGroup
