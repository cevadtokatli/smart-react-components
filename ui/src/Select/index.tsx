import Div from '@smart-react-components/core/Element/Div'
import { IntrinsicStyledProps } from '@smart-react-components/core/element-props/intrinsic-styled-props'
import { ContentElement, PaletteProp, Partial, ResponsiveProp, SetState, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import FormBlockLabel from '../components/FormBlockLabel'
import useInputAddons from '../hooks/useInputAddons'
import { FormValue } from '../types'
import SelectElement from './SelectElement'
import HiddenInput from '../components/HiddenInput'

export interface Props extends
  Partial<ResponsiveProp<'size', SizeProp>>,
  IntrinsicStyledProps {
  active: FormValue | FormValue[]
  children: JSX.Element | JSX.Element[]
  hasBorder?: boolean
  hasHover?: boolean
  hasWaveEffect?: boolean
  isBlock?: boolean
  isDisabled?: boolean
  isRequired?: boolean
  isOutline?: boolean
  isSoft?: boolean
  label?: ContentElement
  leftAddon?: ContentElement
  palette?: PaletteProp
  rightAddon?: ContentElement
  shape?: ShapeProp
  setActive: SetState<FormValue | FormValue[]>
  waveEffectPalette?: PaletteProp
}

const Select = React.forwardRef<HTMLElement, Props>((props, forwardRef) => {
  const { leftAddon, rightAddon } = useInputAddons({
    hasBorder: props.hasBorder,
    isDisabled: props.isDisabled,
    isFocused: false,
    isOutline: props.isOutline,
    isSoft: props.isSoft,
    leftAddon: props.leftAddon,
    palette: props.palette,
    rightAddon: props.rightAddon,
    shape: props.shape,
    size: props.size,
    sizeSm: props.sizeSm,
    sizeMd: props.sizeMd,
    sizeLg: props.sizeLg,
    sizeXl: props.sizeXl,
  })

  const getInputValue = () => (
    (!Array.isArray(props.active) && typeof props.active !== 'undefined' && props.active !== null)
    || (Array.isArray(props.active) && props.active.length > 0)
  )
    ? String(props.active)
    : ''

  return (
    <FormBlockLabel
      formSize={props.size}
      formSizeSm={props.sizeSm}
      formSizeMd={props.sizeMd}
      formSizeLg={props.sizeLg}
      formSizeXl={props.sizeXl}
      isBlock={props.isBlock}
      isDisabled={props.isDisabled}
    >
      { props.label && <span>{props.label}</span> }
      <Div
        display="flex"
        flex="1 1 auto"
      >
        { leftAddon && leftAddon }
        <SelectElement
          hasBorder={props.hasBorder}
          hasLeftAddon={!!leftAddon}
          hasRightAddon={!!rightAddon}
          hasSeparatedLeftAddon={leftAddon?.props?.isSeparated}
          hasSeparatedRightAddon={rightAddon?.props?.isSeparated}
          inputSize={props.size}
          inputSizeSm={props.sizeSm}
          inputSizeMd={props.sizeMd}
          inputSizeLg={props.sizeLg}
          inputSizeXl={props.sizeXl}
          isDisabled={props.isDisabled}
          isFocused={false}
          isOutline={props.isOutline}
          isSoft={props.isSoft}
          palette={props.palette}
          shape={props.shape}
        >
          <HiddenInput
            onChange={() => {}}
            ref={forwardRef}
            value={getInputValue()}
            {...(props.isDisabled && { disabled: true })}
            {...(props.isRequired && { required: true })}
          />
          { (Array.isArray(props.children) ? props.children : [props.children]).map((item, idx) => React.cloneElement(item, {
            key: item.key ?? idx,
            active: props.active,
            hasHover: props.hasHover,
            hasWaveEffect: props.hasWaveEffect,
            isEmbedded: true,
            isOutline: props.isOutline,
            isSoft: props.isSoft,
            palette: props.palette,
            setActive: props.setActive,
            waveEffectPalette: props.waveEffectPalette,
            ...(props.isDisabled && { isDisabled: true }),
          })) }
        </SelectElement>
        { rightAddon && rightAddon }
      </Div>
    </FormBlockLabel>
  )
})

Select.defaultProps = {
  hasBorder: true,
  hasHover: true,
  hasWaveEffect: true,
  isBlock: true,
  isOutline: true,
  palette: 'primary',
  shape: 'rectangle',
  size: 'medium',
  waveEffectPalette: 'light',
}

export default Select
