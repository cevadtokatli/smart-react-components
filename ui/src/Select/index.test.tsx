import { wrapTheme } from '@smart-react-components/core/test'
import { render } from '@testing-library/react'
import React from 'react'
import Select from './'
import Option from './Option'
import OptionGroup from './OptionGroup'

describe('<Select />', () => {
  it('should render component', () => {
    const { asFragment } = render(wrapTheme(
      <Select
        active="item-1"
        setActive={jest.fn()}
      >
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </Select>
    ))
    expect(asFragment()).toMatchSnapshot()
  })
})
