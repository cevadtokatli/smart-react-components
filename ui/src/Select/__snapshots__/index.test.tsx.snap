// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render component 1`] = `
<DocumentFragment>
  .c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

.c6 {
  font-weight: 500;
}

.c7 {
  isolation: isolate;
  overflow: hidden;
  position: relative;
  -webkit-tap-highlight-color: transparent;
}

.c0 > span {
  margin-bottom: .5rem;
}

.c0 > span {
  font-size: 1rem;
}

.c1 > span {
  cursor: default;
  display: block;
  word-break: break-word;
  word-wrap: break-word;
  -webkit-tap-highlight-color: transparent;
}

.c3 {
  font-size: .875rem;
}

.c3 > div {
  margin: 0.225rem 0.45rem;
  padding: 0.225rem 0.45rem;
}

.c4 {
  border: solid 1px;
  border-radius: .2rem;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  font-family: inherit;
  min-width: 1px;
  -webkit-transition: 200ms 0s ease-in-out;
  transition: 200ms 0s ease-in-out;
  -webkit-transition-property: background,border,color,fill;
  transition-property: background,border,color,fill;
  -webkit-appearance: none;
  -webkit-tap-highlight-color: transparent;
  background: #fff;
  border-color: #dbdbdb;
  color: #323a46;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
}

.c4:active,
.c4:focus {
  outline: 0;
}

.c5 {
  height: 100%;
  left: 0;
  opacity: 0;
  pointer-events: none;
  position: absolute;
  top: 0;
  width: 100%;
}

.c8 {
  border-radius: .2rem;
  cursor: default;
  -webkit-transition: 100ms 0s ease-in-out;
  transition: 100ms 0s ease-in-out;
  -webkit-transition-property: background,color,fill;
  transition-property: background,color,fill;
  background: #0553AC;
  color: #fff;
}

.c9 {
  border-radius: .2rem;
  cursor: default;
  -webkit-transition: 100ms 0s ease-in-out;
  transition: 100ms 0s ease-in-out;
  -webkit-transition-property: background,color,fill;
  transition-property: background,color,fill;
}

.c9:hover {
  background: #0665d0;
  color: #fff;
}

<div
    class="c0 c1"
  >
    <div
      class="c2"
      display="flex"
    >
      <div
        class="c3 c4"
        shape="rectangle"
      >
        <input
          class="c5"
          value="item-1"
        />
        <div
          class="c6"
          font-weight="$fontWeight.semibold"
        >
          Category
        </div>
        <div
          class="c7 c8"
          overflow="hidden"
        >
          Item 1
        </div>
        <div
          class="c7 c9"
          overflow="hidden"
        >
          Item 2
        </div>
        <div
          class="c7 c9"
          overflow="hidden"
        >
          Item 3
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
